src/webgpu/api/validation/encoding/cmds/debug.spec.ts

```diff
@@ -0,0 +1,143 @@
+      case 'render pass':
+        this.#commandEncoder = this.device.createCommandEncoder({});
+        return this.#commandEncoder.beginRenderPass({
+          colorAttachments: [
+            {
+              attachment: this.makeAttachmentTexture().createView(),
+              loadValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
+            },
+          ],
+        });
+    }
+  }
+
+  finishEncoder(encoder: Encoders, encoderType: EncoderTypes) {
+    let commandBuffer: GPUCommandBuffer | undefined = undefined;
+    switch (encoderType) {
+      case 'non-pass': {
+        commandBuffer = (encoder as GPUCommandEncoder).finish({});
+        break;
+      }
+      case 'render bundle': {
+        const bundle = (encoder as GPURenderBundleEncoder).finish({});
+        const commandEncoder = this.device.createCommandEncoder({});
+        const pass = commandEncoder.beginRenderPass({
+          colorAttachments: [
+            {
+              attachment: this.makeAttachmentTexture().createView(),
+              loadValue: { r: 1.0, g: 0.0, b: 0.0, a: 1.0 },
+            },
+          ],
+        });
+        pass.executeBundles([bundle]);
+        pass.endPass();
+        commandBuffer = commandEncoder.finish({});
+        break;
+      }
+      case 'compute pass':
+      case 'render pass': {
+        assert(this.#commandEncoder !== undefined);
+        (encoder as GPUComputePassEncoder | GPURenderPassEncoder).endPass();
+        commandBuffer = this.#commandEncoder?.finish();
+        this.#commandEncoder = undefined;
+        break;
+      }
+    }
+    if (commandBuffer !== undefined) {
+      this.queue.submit([commandBuffer]);
+    }
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('debug_group_balanced')
+  .params(
+    params()
+      .combine(poptions('encoderType', kEncoderTypes))
+      .combine(poptions('pushCount', [0, 1, 2]))
+      .combine(poptions('popCount', [0, 1, 2]))
+  )
+  .fn(t => {
+    const encoder = t.createEncoder(t.params.encoderType);
+    const shouldError = t.params.popCount !== t.params.pushCount;
+    t.expectValidationError(() => {
+      for (let i = 0; i < t.params.pushCount; ++i) {
+        encoder.pushDebugGroup(`${i}`);
+      }
+      for (let i = 0; i < t.params.popCount; ++i) {
+        encoder.popDebugGroup();
+      }
+      t.finishEncoder(encoder, t.params.encoderType);
+    }, shouldError);
+  });
+
+g.test('debug_group_empty_string')
+  .params(poptions('encoderType', kEncoderTypes))
+  .fn(t => {
+    const encoder = t.createEncoder(t.params.encoderType);
+    encoder.pushDebugGroup('');
+    encoder.popDebugGroup();
+    t.finishEncoder(encoder, t.params.encoderType);
+  });
+
+g.test('debug_marker_empty_string')
+  .params(poptions('encoderType', kEncoderTypes))
+  .fn(t => {
+    const encoder = t.createEncoder(t.params.encoderType);
+    encoder.insertDebugMarker('');
+    t.finishEncoder(encoder, t.params.encoderType);
+  });
+
+g.test('debug_marker_non_empty_string')
```

Comment:
tiny nit: These two tests could be one, with a `.params(poptions('label', ['', 'marker']))`

Similarly, could expand `debug_group_empty_string` in the same way (even though I know that case is already hit by the previous test).